sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/rel/BeamJoinRel.java

```diff
@@ -194,6 +203,28 @@ public BeamJoinRel(RelOptCluster cluster, RelTraitSet traits, RelNode left, RelN
     }
   }
 
+  private void verifySupportedTrigger(PCollection<Row> pCollection) {
+    WindowingStrategy windowingStrategy = pCollection.getWindowingStrategy();
+
+    if (UNBOUNDED.equals(pCollection.isBounded())
+        && !triggersOncePerWindow(windowingStrategy)) {
+      throw new UnsupportedOperationException(
+          "Joining unbounded PCollections is currently only supported for "
+              + "non-global windows with triggers that are known to produce output once per window,"
+              + "such as the default trigger with zero allowed lateness. "
+              + "In these cases Beam can guarantee it joins all input elements once per window. "
+              + windowingStrategy + " is not supported");
+    }
+  }
+
+  private boolean triggersOncePerWindow(WindowingStrategy windowingStrategy) {
+    Trigger trigger = windowingStrategy.getTrigger();
+
+    return !(windowingStrategy.getWindowFn() instanceof GlobalWindows)
+        && trigger instanceof DefaultTrigger
```

Comment:
From [JavaDoc](https://beam.apache.org/documentation/sdks/javadoc/2.2.0/org/apache/beam/sdk/transforms/windowing/DefaultTrigger.html), it's `equivalent to Repeatedly.forever(AfterWatermark.pastEndOfWindow())`, is it safe for `trigger instanceof DefaultTrigger` as I don't see a subclass for `DefaultTrigger`.